synchronous-proxy/ui/main.go

```diff
@@ -0,0 +1,108 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	synchronousproxy "github.com/temporalio/samples-go/synchronous-proxy"
+	"go.temporal.io/sdk/client"
+)
+
+func main() {
+	c, err := client.NewClient(client.Options{})
+	if err != nil {
+		log.Fatalln("Unable to create client", err)
+	}
+	defer c.Close()
+
+	fmt.Println("T-Shirt Order")
+
+	status, err := CreateOrder(c)
+	if err != nil {
+		log.Fatalln("Unable to create order", err)
+	}
+
+	for {
+		email := PromptAndReadInput("Please enter you email address:")
+		status, err = UpdateOrder(c, status.OrderID, synchronousproxy.RegisterStage, email)
+		if err != nil {
+			log.Println("invalid email", err)
+			continue
+		}
+
+		break
+	}
+
+	for {
+		size := PromptAndReadInput("Please enter your requested size:")
+		status, err = UpdateOrder(c, status.OrderID, synchronousproxy.SizeStage, size)
+		if err != nil {
+			log.Println("invalid size", err)
+			continue
+		}
+
+		break
+	}
+
+	for {
+		color := PromptAndReadInput("Please enter your required tshirt color:")
+		status, err = UpdateOrder(c, status.OrderID, synchronousproxy.ColorStage, color)
+		if err != nil {
+			log.Println("invalid color", err)
+			continue
+		}
+
+		break
+	}
+
+	fmt.Println("Thanks for your order!")
+	fmt.Println("You will receive an email with shipping details shortly")
+}
+
+func PromptAndReadInput(prompt string) string {
+	fmt.Println(prompt)
+
+	reader := bufio.NewReader(os.Stdin)
+	text, _ := reader.ReadString('\n')
+	return strings.TrimSpace(text)
+}
+
+func CreateOrder(c client.Client) (synchronousproxy.OrderStatus, error) {
```

Comment:
I'm not sure that duplicate orders are a problem, they would just time out. I think thats fine. Passing order ID requires creating order ID in the UI which feels unlikely to be the general case.

Reply:
I agree, and maybe even put a relatively human timeout on the workflow as a reminder. IRL unfinished orders would prompt the user via mail after x hours and a after a day or two go away a.k.a time out.

Reply:
Duplicate order would not be a good experience. As in your example, user might already completed their order but the duplicate one could send reminder email after x hour which is annoying. In your, it could be as simple as a UUID created by caller and pass in, this would prevent the caller's retry from creating duplicate workflows.

Reply:
The only place duplicates can be created is the `CreateOrder` call, at which point no email address is registered, so we can't email them when it times out.